subsys/net/lib/lwm2m/lwm2m_engine.c

```diff
@@ -1450,28 +1455,104 @@ int lwm2m_init_message(struct lwm2m_message *msg)
 	return r;
 }
 
+int lwm2m_open_socket(struct lwm2m_ctx *client_ctx)
+{
+	if (client_ctx->sock_fd < 0) {
+		/* open socket */
+#if defined(CONFIG_LWM2M_DTLS_SUPPORT)
+		if (client_ctx->use_dtls) {
+			client_ctx->sock_fd = socket(client_ctx->remote_addr.sa_family, SOCK_DGRAM,
+						     IPPROTO_DTLS_1_2);
+		} else
+#endif /* CONFIG_LWM2M_DTLS_SUPPORT */
+		{
+			client_ctx->sock_fd =
+				socket(client_ctx->remote_addr.sa_family, SOCK_DGRAM, IPPROTO_UDP);
+		}
+
+		if (client_ctx->sock_fd < 0) {
+			LOG_ERR("Failed to create socket: %d", errno);
+			return -errno;
+		}
+		
+		lwm2m_socket_update(client_ctx);		
+	}
+	
+	return 0;
+}
+
+int lwm2m_close_socket(struct lwm2m_ctx *client_ctx)
+{
+	int ret = 0;
+
+	if (client_ctx->sock_fd >= 0) {
+		ret = close(client_ctx->sock_fd);
+		if (ret) {
+			LOG_ERR("Failed to close socket: %d", errno);
+			ret = -errno;
+			return ret;
+		}
+
+		client_ctx->sock_fd = -1;
+		client_ctx->connection_suspended = true;
 #if defined(CONFIG_LWM2M_QUEUE_MODE_ENABLED)
-int lwm2m_engine_connection_resume(struct lwm2m_ctx *client_ctx)
+		/* Enable Queue mode buffer store */
+		client_ctx->buffer_client_messages = true;
+#endif
+		lwm2m_socket_update(client_ctx);
+	}
+	return ret;
+}
+
+int lwm2m_close_suspend(struct lwm2m_ctx *client_ctx)
```

Comment:
Why `close_suspend`? Nothing seems to be closed in this case. Perhaps `lwm2m_socket_suspend()` would be enough?